File: backend/admin_api/serializers.py
from rest_framework import serializers
from django.contrib.auth import get_user_model
from django.utils import timezone
from django.db.models import Sum as SumModel

from admin_api.models import (
    SystemSetting, FeatureFlag, AppVersion,
    SupportTicket, SupportReply, AuditLog, AdminSession,
)
from accounts.models import User
from customers.models import CustomerProfile
from licenses.models import License, LicensePlan, LicenseAuditLog
from payments.models import Payment, Invoice
from notifications.models import Notification

User = get_user_model()


class AdminUserListSerializer(serializers.ModelSerializer):
    class Meta:
        model = User
        fields = [
            'id', 'email', 'username', 'phone', 'first_name', 'last_name',
            'is_active', 'is_admin', 'is_superuser', 'date_joined',
            'last_login', 'created_at', 'updated_at',
        ]
        read_only_fields = ['id', 'date_joined', 'last_login', 'created_at', 'updated_at']


class AdminUserCreateSerializer(serializers.ModelSerializer):
    password = serializers.CharField(write_only=True, min_length=8)
    password2 = serializers.CharField(write_only=True, min_length=8)

    class Meta:
        model = User
        fields = [
            'username', 'email', 'password', 'password2', 'phone',
            'first_name', 'last_name', 'is_admin', 'is_superuser',
        ]

    def validate(self, attrs):
        if attrs['password'] != attrs.pop('password2'):
            raise serializers.ValidationError({'password2': 'Passwords do not match.'})
        return attrs

    def create(self, validated_data):
        password = validated_data.pop('password')
        user = User(**validated_data)
        user.set_password(password)
        user.is_staff = True
        user.save()
        return user


class AdminUserDetailSerializer(serializers.ModelSerializer):
    class Meta:
        model = User
        exclude = ['password', 'user_permissions', 'groups']
        read_only_fields = ['id', 'date_joined', 'last_login', 'created_at', 'updated_at']


class AdminUserUpdateSerializer(serializers.ModelSerializer):
    class Meta:
        model = User
        fields = [
            'email', 'phone', 'first_name', 'last_name',
            'is_active', 'is_admin', 'is_superuser',
        ]


class BusinessListSerializer(serializers.ModelSerializer):
    business_name = serializers.CharField(source='get_full_name', read_only=True)
    company_name = serializers.SerializerMethodField()
    customer_profile_id = serializers.SerializerMethodField()
    customer_status = serializers.SerializerMethodField()
    license_count = serializers.SerializerMethodField()
    total_paid = serializers.SerializerMethodField()
    current_plan = serializers.SerializerMethodField()
    license_status = serializers.SerializerMethodField()
    license_expiry = serializers.SerializerMethodField()

    class Meta:
        model = User
        fields = [
            'id', 'email', 'username', 'phone', 'business_name', 'company_name',
            'business_type', 'is_active', 'is_customer',
            'email_verified', 'phone_verified',
            'customer_profile_id', 'customer_status',
            'license_count', 'total_paid',
            'current_plan', 'license_status', 'license_expiry',
            'date_joined', 'last_login', 'created_at', 'updated_at',
            'notes',
        ]
        read_only_fields = fields

    def get_customer_profile_id(self, obj):
        profile = getattr(obj, 'customer_profile', None)
        return profile.id if profile else None

    def get_customer_status(self, obj):
        profile = getattr(obj, 'customer_profile', None)
        return profile.status if profile else None

    def get_company_name(self, obj):
        profile = getattr(obj, 'customer_profile', None)
        return profile.company_name if profile and profile.company_name else (obj.business_name or obj.get_full_name())

    def get_current_plan(self, obj):
        license = next(iter(obj.licenses.all()), None)
        return license.plan.name if license and license.plan else None

    def get_license_status(self, obj):
        license = next(iter(obj.licenses.all()), None)
        return license.status if license else None

    def get_license_expiry(self, obj):
        license = next(iter(obj.licenses.all()), None)
        return license.expiry_date.isoformat() if license and license.expiry_date else None

    def get_license_count(self, obj):
        return obj.licenses.count()

    def get_total_paid(self, obj):
        total = obj.payments.filter(status='approved').aggregate(
            total=SumModel('amount')
        )['total']
        return float(total) if total else 0.0


class BusinessDetailSerializer(serializers.ModelSerializer):
    customer_profile = serializers.SerializerMethodField()
    licenses = serializers.SerializerMethodField()
    recent_payments = serializers.SerializerMethodField()

    class Meta:
        model = User
        fields = [
            'id', 'email', 'username', 'phone', 'first_name', 'last_name',
            'business_name', 'business_type', 'address',
            'is_active', 'is_customer', 'is_admin',
            'email_verified', 'phone_verified',
            'customer_profile', 'licenses', 'recent_payments',
            'date_joined', 'last_login', 'created_at', 'updated_at',
            'notes',
        ]
        read_only_fields = fields

    def get_customer_profile(self, obj):
        profile = getattr(obj, 'customer_profile', None)
        if profile:
            from customers.serializers import CustomerProfileSerializer
            return CustomerProfileSerializer(profile).data
        return None

    def get_licenses(self, obj):
        licenses = obj.licenses.select_related('plan').all()
        from licenses.serializers import LicenseSerializer
        return LicenseSerializer(licenses, many=True).data

    def get_recent_payments(self, obj):
        payments = obj.payments.select_related('plan').order_by('-created_at')[:5]
        from payments.serializers import PaymentSerializer
        return PaymentSerializer(payments, many=True).data


class SubscriptionListSerializer(serializers.ModelSerializer):
    customer_name = serializers.SerializerMethodField()
    customer_email = serializers.EmailField(source='customer.email')
    plan_name = serializers.CharField(source='plan.name')
    plan_price = serializers.DecimalField(source='plan.price', max_digits=10, decimal_places=2)
    days_remaining = serializers.SerializerMethodField()
    platform = serializers.SerializerMethodField()

    class Meta:
        model = License
        fields = [
            'id', 'license_key', 'customer', 'customer_name', 'customer_email',
            'plan', 'plan_name', 'plan_price', 'platform',
            'status', 'start_date', 'expiry_date', 'device_limit',
            'is_trial', 'days_remaining', 'notes',
            'created_at', 'updated_at',
        ]
        read_only_fields = fields

    def get_customer_name(self, obj):
        return obj.customer.get_full_name() or obj.customer.email

    def get_days_remaining(self, obj):
        if obj.expiry_date:
            delta = obj.expiry_date - timezone.now().date()
            return max(delta.days, 0)
        return 0

    def get_platform(self, obj):
        active = [d for d in obj.device_activations.all() if d.is_active]
        return active[0].operating_system if active else None


class SubscriptionDetailSerializer(serializers.ModelSerializer):
    customer_name = serializers.SerializerMethodField()
    customer_email = serializers.EmailField(source='customer.email')
    customer_phone = serializers.CharField(source='customer.phone')
    plan_name = serializers.CharField(source='plan.name')
    plan_details = serializers.SerializerMethodField()
    payments = serializers.SerializerMethodField()
    audit_logs = serializers.SerializerMethodField()
    device_count = serializers.SerializerMethodField()

    class Meta:
        model = License
        fields = [
            'id', 'license_key', 'customer', 'customer_name', 'customer_email', 'customer_phone',
            'plan', 'plan_name', 'plan_details',
            'status', 'start_date', 'expiry_date', 'device_limit',
            'is_trial', 'notes', 'payments', 'audit_logs', 'device_count',
            'created_at', 'updated_at',
        ]
        read_only_fields = fields

    def get_customer_name(self, obj):
        return obj.customer.get_full_name() or obj.customer.email

    def get_plan_details(self, obj):
        from licenses.serializers import LicensePlanSerializer
        return LicensePlanSerializer(obj.plan).data if obj.plan else None

    def get_payments(self, obj):
        payments = obj.payments.select_related('plan').all()
        from payments.serializers import PaymentSerializer
        return PaymentSerializer(payments, many=True).data

    def get_audit_logs(self, obj):
        logs = obj.audit_logs.all()[:20]
        from licenses.serializers import LicenseAuditLogSerializer
        return LicenseAuditLogSerializer(logs, many=True).data

    def get_device_count(self, obj):
        return obj.device_activations.filter(is_active=True).count()


class PaymentListSerializer(serializers.ModelSerializer):
    customer_name = serializers.SerializerMethodField()
    customer_email = serializers.EmailField(source='customer.email')
    plan_name = serializers.SerializerMethodField()

    class Meta:
        model = Payment
        fields = [
            'id', 'customer', 'customer_name', 'customer_email',
            'license', 'plan', 'plan_name',
            'amount', 'transaction_id', 'receipt_image',
            'payment_method', 'status', 'admin_notes',
            'reviewed_by', 'reviewed_at',
            'created_at', 'updated_at',
        ]
        read_only_fields = fields

    def get_customer_name(self, obj):
        return obj.customer.get_full_name() or obj.customer.email

    def get_plan_name(self, obj):
        return obj.plan.name if obj.plan else None


class PaymentDetailSerializer(serializers.ModelSerializer):
    customer_name = serializers.SerializerMethodField()
    customer_email = serializers.EmailField(source='customer.email')
    customer_phone = serializers.CharField(source='customer.phone')
    plan_name = serializers.SerializerMethodField()
    plan_details = serializers.SerializerMethodField()
    license_details = serializers.SerializerMethodField()
    invoice = serializers.SerializerMethodField()

    class Meta:
        model = Payment
        fields = [
            'id', 'customer', 'customer_name', 'customer_email', 'customer_phone',
            'license', 'license_details', 'plan', 'plan_name', 'plan_details',
            'amount', 'transaction_id', 'receipt_image',
            'payment_method', 'status', 'admin_notes',
            'reviewed_by', 'reviewed_at', 'invoice',
            'created_at', 'updated_at',
        ]
        read_only_fields = fields

    def get_customer_name(self, obj):
        return obj.customer.get_full_name() or obj.customer.email

    def get_plan_name(self, obj):
        return obj.plan.name if obj.plan else None

    def get_plan_details(self, obj):
        from licenses.serializers import LicensePlanSerializer
        return LicensePlanSerializer(obj.plan).data if obj.plan else None

    def get_license_details(self, obj):
        if obj.license:
            from licenses.serializers import LicenseSerializer
            return LicenseSerializer(obj.license).data
        return None

    def get_invoice(self, obj):
        if hasattr(obj, 'invoice') and obj.invoice:
            from payments.serializers import InvoiceSerializer
            return InvoiceSerializer(obj.invoice).data
        return None


class SystemSettingSerializer(serializers.ModelSerializer):
    class Meta:
        model = SystemSetting
        fields = '__all__'
        read_only_fields = ['id', 'created_at', 'updated_at']


class FeatureFlagSerializer(serializers.ModelSerializer):
    class Meta:
        model = FeatureFlag
        fields = '__all__'
        read_only_fields = ['created_at', 'updated_at']


class AppVersionSerializer(serializers.ModelSerializer):
    class Meta:
        model = AppVersion
        fields = '__all__'
        read_only_fields = ['created_at']


class SupportTicketListSerializer(serializers.ModelSerializer):
    business_name = serializers.SerializerMethodField()
    business_email = serializers.EmailField(source='business.email')
    assigned_to_name = serializers.SerializerMethodField()
    reply_count = serializers.SerializerMethodField()

    class Meta:
        model = SupportTicket
        fields = [
            'id', 'business', 'business_name', 'business_email',
            'subject', 'priority', 'status',
            'assigned_to', 'assigned_to_name',
            'platform', 'reply_count',
            'created_at', 'updated_at',
        ]
        read_only_fields = fields

    def get_business_name(self, obj):
        return obj.business.get_full_name() or obj.business.email

    def get_assigned_to_name(self, obj):
        if obj.assigned_to:
            return obj.assigned_to.get_full_name() or obj.assigned_to.email
        return None

    def get_reply_count(self, obj):
        return obj.replies.count()


class SupportTicketDetailSerializer(serializers.ModelSerializer):
    business_name = serializers.SerializerMethodField()
    business_email = serializers.EmailField(source='business.email')
    business_phone = serializers.CharField(source='business.phone')
    assigned_to_name = serializers.SerializerMethodField()
    replies = serializers.SerializerMethodField()

    class Meta:
        model = SupportTicket
        fields = [
            'id', 'business', 'business_name', 'business_email', 'business_phone',
            'subject', 'description', 'priority', 'status',
            'assigned_to', 'assigned_to_name',
            'platform', 'replies',
            'created_at', 'updated_at',
        ]
        read_only_fields = fields

    def get_business_name(self, obj):
        return obj.business.get_full_name() or obj.business.email

    def get_assigned_to_name(self, obj):
        if obj.assigned_to:
            return obj.assigned_to.get_full_name() or obj.assigned_to.email
        return None

    def get_replies(self, obj):
        replies = obj.replies.select_related('admin').all()
        return SupportReplySerializer(replies, many=True).data


class SupportReplySerializer(serializers.ModelSerializer):
    admin_name = serializers.SerializerMethodField()

    class Meta:
        model = SupportReply
        fields = [
            'id', 'ticket', 'admin', 'admin_name',
            'message', 'is_internal',
            'created_at',
        ]
        read_only_fields = ['id', 'admin', 'admin_name', 'created_at']

    def get_admin_name(self, obj):
        if obj.admin:
            return obj.admin.get_full_name() or obj.admin.email
        return None


class AuditLogSerializer(serializers.ModelSerializer):
    admin_name = serializers.SerializerMethodField()
    admin_email = serializers.EmailField(source='admin.email', read_only=True)

    class Meta:
        model = AuditLog
        fields = '__all__'
        read_only_fields = ['created_at']

    def get_admin_name(self, obj):
        if obj.admin:
            return obj.admin.get_full_name() or obj.admin.email
        return None


class AdminSessionSerializer(serializers.ModelSerializer):
    admin_name = serializers.SerializerMethodField()

    class Meta:
        model = AdminSession
        fields = '__all__'
        read_only_fields = ['login_time']

    def get_admin_name(self, obj):
        if obj.admin:
            return obj.admin.get_full_name() or obj.admin.email
        return None


class NotificationBroadcastSerializer(serializers.Serializer):
    notification_type = serializers.ChoiceField(choices=Notification.Type.choices)
    title = serializers.CharField(max_length=255)
    message = serializers.CharField()
    link = serializers.CharField(max_length=500, required=False, allow_blank=True)
    send_to_all = serializers.BooleanField(default=False)
    recipient_ids = serializers.ListField(
        child=serializers.IntegerField(),
        required=False,
        default=list,
    )

    def validate(self, attrs):
        if not attrs.get('send_to_all') and not attrs.get('recipient_ids'):
            raise serializers.ValidationError(
                "Either send_to_all=True or recipient_ids must be provided."
            )
        return attrs


class ExtendSubscriptionSerializer(serializers.Serializer):
    days = serializers.IntegerField(min_value=1, required=True)


class UpgradeDowngradeSerializer(serializers.Serializer):
    plan_id = serializers.PrimaryKeyRelatedField(
        queryset=LicensePlan.objects.filter(is_active=True),
        required=True,
    )


class SubscriptionNotesSerializer(serializers.Serializer):
    notes = serializers.CharField(required=True)


class PaymentApproveSerializer(serializers.Serializer):
    admin_notes = serializers.CharField(required=False, allow_blank=True, default='')


class PaymentRejectSerializer(serializers.Serializer):
    reason = serializers.CharField(required=True)


class PaymentRequestInfoSerializer(serializers.Serializer):
    message = serializers.CharField(required=True)


class ExtendTrialSerializer(serializers.Serializer):
    days = serializers.IntegerField(min_value=1, required=True)


class ConvertTrialSerializer(serializers.Serializer):
    plan_id = serializers.PrimaryKeyRelatedField(
        queryset=LicensePlan.objects.filter(is_active=True),
        required=True,
    )


class SupportTicketCreateSerializer(serializers.ModelSerializer):
    class Meta:
        model = SupportTicket
        fields = ['business', 'subject', 'description', 'priority', 'platform']


class SupportReplyCreateSerializer(serializers.Serializer):
    message = serializers.CharField(required=True)
    is_internal = serializers.BooleanField(default=False)


class SupportTicketAssignSerializer(serializers.Serializer):
    assigned_to = serializers.PrimaryKeyRelatedField(
        queryset=User.objects.filter(is_admin=True),
        required=True,
    )


class SupportTicketNotesSerializer(serializers.Serializer):
    notes = serializers.CharField(required=True)


class AppVersionNotifySerializer(serializers.Serializer):
    message = serializers.CharField(required=False, default='')

File: api/src/lib/serializers.ts
import type {
  User,
  Plan,
  License,
  DeviceActivation,
  Payment,
  Invoice,
  AuditLog,
  AdminSession,
  AppVersion,
  SystemSetting,
  Notification,
  LicenseAuditLog,
} from "@prisma/client";

/**
 * DRF-compatible serializers. Field names/values mirror rest_framework output
 * so the existing Next.js admin panel keeps working unchanged.
 */

export type UserLike = Pick<
  User,
  | "id"
  | "username"
  | "email"
  | "first_name"
  | "last_name"
  | "phone"
  | "business_name"
  | "business_type"
  | "address"
  | "is_staff"
  | "is_active"
  | "is_superuser"
  | "is_admin"
  | "is_customer"
  | "email_verified"
  | "phone_verified"
  | "date_joined"
  | "last_login"
  | "created_at"
  | "updated_at"
  | "notes"
>;

export function fullName(user: { first_name?: string | null; last_name?: string | null; email?: string | null; username?: string | null } | null | undefined): string {
  if (!user) return "";
  const first = user.first_name || "";
  const last = user.last_name || "";
  const name = [first, last].filter(Boolean).join(" ").trim();
  return name || user.email || user.username || "";
}

/** Matches Django UserSerializer (all fields except password/groups/user_permissions). */
export function serializeUser(user: UserLike): Record<string, unknown> {
  return {
    id: user.id,
    username: user.username,
    email: user.email,
    first_name: user.first_name,
    last_name: user.last_name,
    full_name: fullName(user),
    phone: user.phone,
    business_name: user.business_name,
    business_type: user.business_type,
    address: user.address,
    is_staff: user.is_staff,
    is_active: user.is_active,
    is_superuser: user.is_superuser,
    is_admin: user.is_admin,
    is_customer: user.is_customer,
    email_verified: user.email_verified,
    phone_verified: user.phone_verified,
    date_joined: isoDateTime(user.date_joined),
    last_login: isoDateTime(user.last_login),
    created_at: isoDateTime(user.created_at),
    updated_at: isoDateTime(user.updated_at),
    notes: user.notes,
  };
}

export function serializePlan(plan: Plan): Record<string, unknown> {
  return {
    id: plan.id,
    name: plan.name,
    display_name: plan.name,
    edition: plan.edition,
    duration_months: plan.duration_months,
    device_limit: plan.device_limit,
    price: toNumber(plan.price),
    included_mobile_devices: plan.included_mobile_devices,
    included_desktop_devices: plan.included_desktop_devices,
    included_businesses: plan.included_businesses,
    addon_mobile_price: toNumber(plan.addon_mobile_price),
    addon_desktop_price: toNumber(plan.addon_desktop_price),
    addon_business_price: toNumber(plan.addon_business_price),
    is_active: plan.is_active,
    created_at: isoDateTime(plan.created_at),
  };
}

export function planLabel(plan: Pick<Plan, "name" | "duration_months"> | null | undefined): string | null {
  if (!plan) return null;
  const plural = plan.duration_months === 1 ? " month" : " months";
  return `${plan.name} (${plan.duration_months}${plural})`;
}

/** Matches licenses.serializers.LicenseSerializer. */
export function serializeLicense(
  license: License & { customer?: UserLike; plan?: Plan | null },
): Record<string, unknown> {
  return {
    id: license.id,
    license_key: license.license_key,
    customer: license.customer_id,
    customer_name: license.customer ? fullName(license.customer) : null,
    plan: license.plan_id,
    plan_name: license.plan?.name ?? null,
    plan_details: license.plan ? serializePlan(license.plan) : null,
    status: license.status,
    start_date: isoDate(license.start_date),
    expiry_date: isoDate(license.expiry_date),
    device_limit: license.device_limit,
    max_mobile_devices: license.max_mobile_devices,
    max_desktop_devices: license.max_desktop_devices,
    max_businesses: license.max_businesses,
    notes: license.notes,
    is_trial: license.is_trial,
    created_at: isoDateTime(license.created_at),
    updated_at: isoDateTime(license.updated_at),
  };
}

export function serializeDeviceActivation(
  da: DeviceActivation & { license?: License | null },
): Record<string, unknown> {
  return {
    id: da.id,
    license: da.license_id,
    license_key: da.license?.license_key ?? null,
    device_id: da.device_id,
    device_name: da.device_name,
    activation_date: isoDateTime(da.activation_date),
    last_seen: isoDateTime(da.last_seen),
    ip_address: da.ip_address,
    operating_system: da.operating_system,
    device_type: da.device_type,
    is_active: da.is_active,
  };
}

export function serializeLicenseAuditLog(
  log: LicenseAuditLog & {
    created_by?: UserLike | null;
    license?: License | null;
  },
): Record<string, unknown> {
  return {
    id: log.id,
    license: log.license_id,
    license_key: log.license?.license_key ?? null,
    action: log.action,
    details: parseJsonObject(log.details),
    ip_address: log.ip_address,
    created_by: log.created_by_id,
    created_by_name: log.created_by ? fullName(log.created_by) : null,
    created_at: isoDateTime(log.created_at),
  };
}

/** Matches admin_api.serializers.SubscriptionListSerializer. */
export function serializeSubscriptionList(
  sub: License & { customer?: UserLike | null; plan?: Plan | null; device_activations?: (DeviceActivation | null)[] | null },
): Record<string, unknown> {
  const active = (sub.device_activations ?? []).filter((d) => d && d.is_active);
  return {
    id: sub.id,
    license_key: sub.license_key,
    customer: sub.customer_id,
    customer_name: sub.customer ? fullName(sub.customer) : null,
    customer_email: sub.customer?.email ?? null,
    plan: sub.plan_id,
    plan_name: sub.plan?.name ?? null,
    plan_label: planLabel(sub.plan),
    plan_price: sub.plan ? toNumber(sub.plan.price) : null,
    platform: active[0]?.operating_system ?? null,
    status: sub.status,
    start_date: isoDate(sub.start_date),
    expiry_date: isoDate(sub.expiry_date),
    device_limit: sub.device_limit,
    max_mobile_devices: sub.max_mobile_devices,
    max_desktop_devices: sub.max_desktop_devices,
    max_businesses: sub.max_businesses,
    is_trial: sub.is_trial,
    days_remaining: daysRemaining(sub.expiry_date),
    notes: sub.notes,
    created_at: isoDateTime(sub.created_at),
    updated_at: isoDateTime(sub.updated_at),
  };
}

/** Matches admin_api.serializers.SubscriptionDetailSerializer. */
export function serializeSubscriptionDetail(
  sub: License & {
    customer?: UserLike | null;
    plan?: Plan | null;
    payments?: Payment[] | null;
    audit_logs?: LicenseAuditLog[] | null;
    device_activations?: DeviceActivation[] | null;
  },
): Record<string, unknown> {
  return {
    ...serializeSubscriptionList(sub),
    customer_phone: sub.customer?.phone ?? null,
    plan_details: sub.plan ? serializePlan(sub.plan) : null,
    payments: (sub.payments ?? []).map((p) => serializePaymentList(p)),
    audit_logs: (sub.audit_logs ?? []).map((l) => serializeLicenseAuditLog(l)),
    device_count: (sub.device_activations ?? []).filter((d) => d.is_active).length,
  };
}

/** Matches admin_api.serializers.PaymentListSerializer. */
export function serializePaymentList(
  pmt: Payment & { customer?: UserLike | null; plan?: Plan | null; license?: License | null },
): Record<string, unknown> {
  return {
    id: pmt.id,
    customer: pmt.customer_id,
    customer_name: pmt.customer ? fullName(pmt.customer) : null,
    customer_email: pmt.customer?.email ?? null,
    business_name: (pmt.customer?.business_name || "").trim() || null,
    license: pmt.license_id,
    plan: pmt.plan_id,
    plan_name: pmt.plan?.name ?? null,
    plan_label: planLabel(pmt.plan),
    payment_type: pmt.payment_type,
    quantity: pmt.quantity,
    description: pmt.description,
    amount: toNumber(pmt.amount),
    transaction_id: pmt.transaction_id,
    receipt_image: pmt.receipt_image ?? null,
    payment_method: pmt.payment_method,
    status: pmt.status,
    admin_notes: pmt.admin_notes,
    reviewed_by: pmt.reviewed_by_id,
    reviewed_at: isoDateTime(pmt.reviewed_at),
    created_at: isoDateTime(pmt.created_at),
    updated_at: isoDateTime(pmt.updated_at),
  };
}

/** Matches admin_api.serializers.PaymentDetailSerializer. */
export function serializePaymentDetail(
  pmt: Payment & {
    customer?: UserLike | null;
    plan?: Plan | null;
    license?: License | null;
    invoice?: Invoice | null;
  },
): Record<string, unknown> {
  return {
    ...serializePaymentList(pmt),
    customer_phone: pmt.customer?.phone ?? null,
    plan_details: pmt.plan ? serializePlan(pmt.plan) : null,
    license_details: pmt.license ? serializeLicense(pmt.license) : null,
    invoice: pmt.invoice ? serializeInvoice(pmt.invoice) : null,
  };
}

export function serializeInvoice(inv: Invoice & { customer?: UserLike | null }): Record<string, unknown> {
  return {
    id: inv.id,
    invoice_number: inv.invoice_number,
    customer: inv.customer_id,
    customer_name: inv.customer ? fullName(inv.customer) : null,
    payment: inv.payment_id,
    license: inv.license_id,
    amount: toNumber(inv.amount),
    status: inv.status,
    due_date: isoDate(inv.due_date),
    paid_at: isoDateTime(inv.paid_at),
    created_at: isoDateTime(inv.created_at),
  };
}

export function serializeAuditLog(log: AuditLog & { admin?: UserLike | null }): Record<string, unknown> {
  return {
    id: log.id,
    admin: log.admin_id,
    admin_name: log.admin ? fullName(log.admin) : null,
    admin_email: log.admin?.email ?? null,
    action: log.action,
    resource_type: log.resource_type,
    resource_id: log.resource_id,
    details: parseJsonObject(log.details),
    before_state: parseJsonObject(log.before_state),
    after_state: parseJsonObject(log.after_state),
    ip_address: log.ip_address,
    user_agent: log.user_agent,
    created_at: isoDateTime(log.created_at),
  };
}

export function serializeAdminSession(s: AdminSession & { admin?: UserLike | null }): Record<string, unknown> {
  return {
    id: s.id,
    admin: s.admin_id,
    admin_name: s.admin ? fullName(s.admin) : null,
    ip_address: s.ip_address,
    user_agent: s.user_agent,
    login_time: isoDateTime(s.login_time),
    logout_time: isoDateTime(s.logout_time),
    is_active: s.is_active,
  };
}

export function serializeAppVersion(v: AppVersion): Record<string, unknown> {
  return {
    id: v.id,
    platform: v.platform,
    version: v.version,
    min_version: v.min_version,
    is_force_update: v.is_force_update,
    release_notes: v.release_notes,
    download_url: v.download_url,
    created_at: isoDateTime(v.created_at),
  };
}

export function serializeSystemSetting(s: SystemSetting): Record<string, unknown> {
  return {
    id: s.id,
    key: s.key,
    value: s.value,
    type: s.type,
    description: s.description,
    created_at: isoDateTime(s.created_at),
    updated_at: isoDateTime(s.updated_at),
  };
}

export function serializeNotification(n: Notification & { recipient?: UserLike | null }): Record<string, unknown> {
  return {
    id: n.id,
    recipient: n.recipient_id,
    notification_type: n.notification_type,
    title: n.title,
    message: n.message,
    is_read: n.is_read,
    link: n.link,
    created_at: isoDateTime(n.created_at),
  };
}

// ─── date/number helpers (DRF-compatible) ────────────────────────────────────

export function isoDateTime(d: Date | null | undefined): string | null {
  if (!d) return null;
  return d.toISOString();
}

export function isoDate(d: Date | null | undefined): string | null {
  if (!d) return null;
  const y = d.getUTCFullYear();
  const m = String(d.getUTCMonth() + 1).padStart(2, "0");
  const day = String(d.getUTCDate()).padStart(2, "0");
  return `${y}-${m}-${day}`;
}

export function toNumber(value: unknown): number {
  if (typeof value === "number") return value;
  if (typeof value === "string") {
    const n = Number.parseFloat(value);
    return Number.isFinite(n) ? n : 0;
  }
  if (typeof value === "object" && value !== null && typeof (value as { toString?: unknown }).toString === "function") {
    const n = Number.parseFloat(String((value as { toString(): string }).toString()));
    return Number.isFinite(n) ? n : 0;
  }
  return 0;
}

export function daysRemaining(expiry: Date | null | undefined): number {
  if (!expiry) return 0;
  const now = new Date();
  const today = new Date(Date.UTC(now.getUTCFullYear(), now.getUTCMonth(), now.getUTCDate()));
  const expiryDay = new Date(Date.UTC(expiry.getUTCFullYear(), expiry.getUTCMonth(), expiry.getUTCDate()));
  return Math.max(Math.floor((expiryDay.getTime() - today.getTime()) / 86400000), 0);
}

export function parseJsonObject(value: unknown): unknown {
  if (typeof value !== "string") return value ?? {};
  try {
    const parsed = JSON.parse(value);
    return parsed ?? {};
  } catch {
    return {};
  }
}